edb/edgeql-parser/src/validation.rs

```diff
@@ -0,0 +1,207 @@
+use std::str::FromStr;
+
+use bigdecimal::num_bigint::ToBigInt;
+use bigdecimal::BigDecimal;
+
+use crate::helpers::{unquote_bytes, unquote_string};
+use crate::position::Pos;
+use crate::tokenizer::{Error, Kind, Token, Tokenizer, Value, MAX_KEYWORD_LENGTH};
+
+/// Applies additional validation to the tokens.
+/// Combines multi-word keywords into single tokens.
+pub struct Validator<'a> {
+    pub inner: Tokenizer<'a>,
+
+    pub(super) peeked: Option<Option<Result<Token<'a>, Error>>>,
+    pub(super) keyword_buf: String,
+}
+
+impl<'a> Iterator for Validator<'a> {
+    type Item = Result<Token<'a>, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut token = match self.next_inner()? {
+            Ok(t) => t,
+            Err(e) => return Some(Err(e)),
+        };
+
+        token.value = match parse_value(&token) {
+            Ok(x) => x,
+            Err(e) => return Some(Err(Error::new(e).with_span(token.span))),
+        };
+
+        if let Some(text) = self.combine_multi_word_keywords(&token) {
+            token.kind = Kind::Keyword;
+            token.text = text.into();
+            self.peeked = None;
+        }
+
+        Some(Ok(token))
+    }
+}
+
+impl<'a> Validator<'a> {
+    pub(super) fn new(inner: Tokenizer<'a>) -> Self {
+        Validator {
+            inner,
+            peeked: None,
+            keyword_buf: String::with_capacity(MAX_KEYWORD_LENGTH),
+        }
+    }
+
+    /// Mimics behavior of [std::iter::Peekable]. We could use that, but it
+    /// hides access to underlying iterator.
+    fn next_inner(&mut self) -> Option<Result<Token<'a>, Error>> {
+        if let Some(peeked) = self.peeked.take() {
+            peeked
+        } else {
+            self.inner.next()
+        }
+    }
+
+    /// Mimics behavior of [std::iter::Peekable]. We could use that, but it
+    /// hides access to underlying iterator.
+    fn peek(&mut self) -> &Option<Result<Token<'a>, Error>> {
+        if self.peeked.is_none() {
```

Comment:
Could this be cleaner with a `match` instead of `is_none`+`unwrap`?
Or does that run into some obscure borrow checker disaster?

Reply:
You could do probably do something like:
```rust
match &self.peeked {
   Some(x) => x,
   None => {
       self.peeked = Some(self.inner.next());
       self.peeked.as_ref().unwrap()
   }
}
```
... but that doesn't seem cleaner to me.